marlowe-runtime-web/server/Language/Marlowe/Runtime/Web/Server/Monad.hs

```diff
@@ -39,14 +42,25 @@ newtype AppM r a = AppM { runAppM :: ReaderT (AppEnv r) Handler a }
     , MonadBase IO
     )
 
-instance MonadCleanup (AppM r) where
-  generalCleanup acquire release action = coerce $ generalCleanup
-    (toTransformers acquire)
-    (\a b -> toTransformers $ release a b)
-    (\a -> toTransformers $ action a)
-    where
-      toTransformers :: AppM r a -> ReaderT (AppEnv r) (ExceptT ServerError IO) a
-      toTransformers = coerce
+instance MonadWith (AppM r) where
+  type WithException (AppM r) = WithException (ExceptT ServerError IO)
+  stateThreadingGeneralWith
+    :: forall a b releaseReturn
+     . GeneralAllocate (AppM r) (WithException (ExceptT ServerError IO)) releaseReturn b a
+    -> (a -> AppM r b)
+    -> AppM r (b, releaseReturn)
+  stateThreadingGeneralWith (GeneralAllocate allocA) go = AppM . ReaderT $ \r -> do
+    let
+      allocA' :: (forall x. ExceptT ServerError IO x -> ExceptT ServerError IO x) -> (ExceptT ServerError IO) (GeneralAllocated (ExceptT ServerError IO) (WithException (ExceptT ServerError IO)) releaseReturn b a)
+      allocA' restore = do
+        let
+          restore' :: forall x. AppM r x -> AppM r x
+          restore' mx = AppM . ReaderT $ coerce . restore . coerce . (runReaderT . runAppM) mx
+        GeneralAllocated a releaseA <- (coerce . runReaderT . runAppM) (allocA restore') r
+        let
+          releaseA' relTy = (coerce . runReaderT . runAppM) (releaseA relTy) r
+        pure $ GeneralAllocated a releaseA'
+    coerce $ stateThreadingGeneralWith (GeneralAllocate allocA') (flip (coerce . runReaderT . runAppM) r . go)
```

Comment:
These `MonadWith` instances are pretty monstrous :fearful: Is there no way to wrap these newtype implementations up in reusable default functions?

Reply:
Yeah it's bad, it's on my list to fix. The deriving newtype/deriving via *should* work I think, there's something funky in GHC's unification.

If that's not fixable, it might be possible to have default implementations using a [ControlTransformation](https://hackage.haskell.org/package/eventuo11y-0.9.0.1/docs/Control-Natural-Control.html#t:ControlTransformation)